mlflow/tracking/fluent.py

```diff
@@ -80,20 +89,48 @@ def set_experiment(experiment_name: str) -> None:
         Tags: {}
         Lifecycle_stage: active
     """
-    client = MlflowClient()
-    experiment = client.get_experiment_by_name(experiment_name)
-    exp_id = experiment.experiment_id if experiment else None
-    if exp_id is None:  # id can be 0
-        print("INFO: '{}' does not exist. Creating a new experiment".format(experiment_name))
-        exp_id = client.create_experiment(experiment_name)
-    elif experiment.lifecycle_stage == LifecycleStage.DELETED:
+    if (experiment_name is not None and experiment_id is not None) or (
+        experiment_name is None and experiment_id is None
+    ):
         raise MlflowException(
-            "Cannot set a deleted experiment '%s' as the active experiment."
-            " You can restore the experiment, or permanently delete the "
-            " experiment to create a new one." % experiment.name
+            message="Must specify exactly one of: `experiment_id` or `experiment_name`.",
+            error_code=INVALID_PARAMETER_VALUE,
         )
+
+    def verify_experiment_active(experiment):
+        if experiment.lifecycle_stage == LifecycleStage.DELETED:
+            raise MlflowException(
+                message=(
+                    "Cannot set a deleted experiment '%s' as the active experiment."
+                    " You can restore the experiment, or permanently delete the "
+                    " experiment to create a new one." % experiment.name
+                ),
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+
+    client = MlflowClient()
+    if experiment_id is None:
+        experiment = client.get_experiment_by_name(experiment_name)
+        if experiment:
+            verify_experiment_active(experiment)
+            experiment_id = experiment.experiment_id
+        else:
+            _logger.info(
+                "Experiment with name '%s' does not exist. Creating a new experiment.",
+                experiment_name,
+            )
+            experiment_id = client.create_experiment(experiment_name)
+    else:
+        experiment = client.get_experiment(experiment_id)
```

Comment:
Does this function throw if experiment_id doesn't exist?

Reply:
404 in REST and throws in file_store and sqlalchemy_store

Reply:
Thanks @BenWilson2 ! Yes, while this throws for MLflow-native stores, it may not always throw for alternative stores. I don't think the `None` check is particularly problematic here as an extra layer of defense.